### DISCUSSION TITLE
Is there a canoncal way for SPEC file to be backward compatible with older rpm versions?

### DISCUSSION THREAD

**Discussion by dletai** [rpm-software-management/rpm#3314] [Q&A]
According to https://rpm-software-management.github.io/rpm/manual/spec.html some sections are only available from certain versions, for example %conf (since rpm >= 4.18).
I'd like my SPEC file to use newer capabilities if rpmbuild supports those, and fallback to older capabilities if built on an older system.

I was hoping for something like
```
%if %rpmversion >= 4.18
%conf
%else
%build
%endif
%configure
%if %rpmversion >= 4.18
%build
%endif
make
```
But I don't see anything like %rpmversion being provided. I can always `rpm -q rpm --qf %{VERSION}` but that has 2 issues
1. I'm not sure if this will evaluates correctly to a version comparison and not be used in a lexographical string comaprison
2. It doesn't seem like the correct method, since rpmbuild should 'know' it's own capabilites (including version).

Does it make sense to build a spec file that will behave differently on fedora vs. rhel ?
Is there a canonical way to enjoy the best capabilities of a given rpm version while maintaining backward compatibility?

**Comment by pmatilai** [comment]
`%{rpmversion}` built-in macro, verbatim to what you ask, exists since rpm >= 4.18.2 :smile: 

I would not recommend conditionalizing too much on it, certainly not spec sections. That'll only turn your specs into unreadable spaghetti with very little in return.

**Reply by dletai** [reply]
I am curious about new feature introduction into SPEC files in that case. SPEC files are immutable, even between distro versions. A new app packaged has to consider being built on both EL7 and EL9 while (10 years) over supported lifetime. Are SPEC files to remain at lowest feature set a decade after new features are introduced? Is there no canonical way to bridge the gap?

It gets worse when considering that 3rd party won't create a different SPEC file for newest fedora and oldest supported EL. The assumption would be to rpmbuild on target distro, but not to have different SPEC for each.

**Reply by pmatilai** [reply]
It's a tradeoff like anything. Having seen my share of specs that attempted to be compatible across multiple distros and a dozen rpm versions... me, I'd rather maintain separate specs. 

Note the above was really quite specific to %conf: the benefit of using %conf over just doing it in %build doesn't generally justify the mess it causes. If it actually is that useful to you, your choice. I'd be interested in hearing that usecase though :smile:

**Reply by dletai** [reply]
I hear you. The benefit is mostly during development of the SPEC, having to wait only on -bf as opposed to -bc is quite noticeable during development and debug, but for the production SPEC I suppose I will drop it.

**Reply by dletai** [reply]
This doesn't solve the issue for rpm before 4.18.2 that won't know how to parse this correctly - basically any EL9 and below won't work correctly with this approach.

**Reply by pmatilai** [reply]
Verbatim `%if %rpmversion >= xxx` wont work anywhere at all, you need to apply the "something like" too. For versions that actually support actual version comparison in expressions (>= 4.16) you can write it like this to make sure it does the right thing if %rpmversion is not defined:

```
%if v"0%{?rpmversion}" >= v"4.18"
...
%endif
```

That wont help with < 4.16 though. It's usually better to test for features rather than versions. And simply using explicit switches at build-time may well be the sanest option because, well, historically rpm does not really support this kind of thing at all.
